File: Daily_Challenge/108_sorted_array_to_binary_tree/ArrayBinaryTree.py


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


def sortedArrayToBST(nums):
    
    def recur(left, right):
        if left > right:
            return None
        mid = (left + right) // 2
        return TreeNode(val = nums[mid], left = recur(left, mid - 1), right = recur(mid + 1, right))
    
    return recur(0, len(nums) - 1)


nums = [-2, -1, 0, 1, 2]
root = sortedArrayToBST(nums)

File: Daily_Challenge/108_sorted_array_to_binary_tree/ArrayBinaryTree.c
#include <stdlib.h>
#include <stdio.h> 
 


struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};


struct TreeNode* recur(int left, int right, int* nums) {
    if (left > right) {
        return NULL;
    }
    int mid = (left + right) / 2;
    struct TreeNode* node;
    node = (struct TreeNode*)malloc(sizeof(struct TreeNode)); 
    node -> val = nums[mid];
    node -> left = recur(left, mid - 1, nums);
    node -> right = recur(mid + 1, right, nums);
    return node;
}

struct TreeNode* sortedArrayToBST(int* nums, int numsSize){
    return recur(0, numsSize - 1, nums);
}

int main() {

    int nums[5] = {-2, -1, 0, 1, 2};
    int numsSize = 5;
    struct TreeNode* root = sortedArrayToBST(nums, numsSize); 
    return 0;
}